### ISSUE TITLE
[xpu-alignment] _adaptive_avg_pool3d_backward on XPU accepts mismatched leading dims and reads gradOutput past its end

### ISSUE THREAD

**Issue by laifenxiawucha** [intel/torch-xpu-ops#5232]
## Independent local verification

XPU rerun 2026-09-07 (Max 1100, torch 2.15.0.dev20260902+xpu): direct aten._adaptive_avg_pool3d_backward accepted 4D channel and 5D batch/channel mismatches and returned input-shaped gradients instead of raising.

The upstream result is still tracked as upstream-dependent; this issue records the concrete XPU defect and does not assert that upstream has finalized cross-backend semantics.

<sub>alignment scan `2026-08-30`, run `33350143612`</sub>
## Summary

Upstream source: pytorch/pytorch#195304 (open) - `Validate leading dims of input and gradOutput in _adaptive_avg_pool3d_backward`, which fixes pytorch/pytorch#194800 (`[CPU] _adaptive_avg_pool3d_backward reads past a smaller grad_output tensor`, labelled `module: crash`).
Alignment scan window: `[2026-08-30T00:00:00Z, 2026-08-31T00:00:00Z)` over `pytorch/pytorch`.

`_adaptive_avg_pool3d_backward` only checked that `input` and `gradOutput` have equal rank. The kernel then derives its batch and channel loop counts from `input` while reading `gradOutput`, so a smaller batch or channel dimension on `gradOutput` makes later iterations read past the end of its buffer. The upstream PR adds a per-dimension size check to the CPU template and, separately, to the CUDA template. XPU has its own kernel and did not receive the check.

## Observed XPU behavior

Every mismatch is accepted and the call returns an input-shaped tensor, exactly the pre-fix signature; the tested build predates the fix, which is why CPU behaves the same way:

```
4D channel mismatch: input=(4, 2, 2, 2) grad_output=(2, 1, 1, 1)
  cpu: returned (4, 2, 2, 2)
  xpu: returned (4, 2, 2, 2)
5D batch mismatch: input=(3, 2, 2, 2, 2) grad_output=(1, 2, 1, 1, 1)
  cpu: returned (3, 2, 2, 2, 2)
  xpu: returned (3, 2, 2, 2, 2)
5D channel mismatch: input=(1, 4, 2, 2, 2) grad_output=(1, 2, 1, 1, 1)
  cpu: returned (1, 4, 2, 2, 2)
  xpu: returned (1, 4, 2, 2, 2)
```

After the upstream fix both CPU and CUDA raise `adaptive_avg_pool3d_backward: input and gradOutput_ must have the same size at dim <i>, but got <a> and <b>` for all three cases.

Scope of the evidence: these runs establish that the guard is absent on XPU and that the call returns instead of raising. They do not demonstrate the out-of-bounds read itself, which would need an ASan-instrumented build.

## Reproducer

```python
import torch

CASES = (
    ("4D channel mismatch", (4, 2, 2, 2), (2, 1, 1, 1)),
    ("5D batch mismatch", (3, 2, 2, 2, 2), (1, 2, 1, 1, 1)),
    ("5D channel mismatch", (1, 4, 2, 2, 2), (1, 2, 1, 1, 1)),
)

for label, input_shape, grad_shape in CASES:
    torch.manual_seed(0)
    x = torch.randn(input_shape, device="xpu")
    grad_output = torch.randn(grad_shape, device="xpu")
    try:
        out = torch.ops.aten._adaptive_avg_pool3d_backward(grad_output, x)
        torch.xpu.synchronize()
        print(label, "returned", tuple(out.shape))
    except RuntimeError as exc:
        print(label, "raised", str(exc).splitlines()[0])
```

## Target path

`_adaptive_avg_pool3d_backward` dispatches per backend (`CPU: adaptive_avg_pool3d_backward_cpu`, `CUDA: adaptive_avg_pool3d_backward_cuda`, `XPU: adaptive_avg_pool3d_backward_xpu`), so there is no shared validation layer.

- `src/ATen/native/xpu/AdaptiveAveragePooling3d.cpp`: `adaptive_avg_pool3d_backward_xpu` and `adaptive_avg_pool3d_backward_out_xpu` add no shape validation of their own.
- `src/ATen/native/xpu/sycl/AdaptiveAveragePooling3dKernels.cpp`: `adaptive_avg_pool3d_backward_kernel` performs `adaptive_pool_empty_output_check` and the rank equality check, then derives `sizeD` from `input.size(0)` or `input.size(1)`, `sizeB` from `input.size(0)` and `totalZ` from `input` sizes, while indexing `gradOutput`. No dimension-by-dimension comparison of the leading dims exists.

The fix mirrors the upstream hunk: for `i` in `[0, input.dim() - 3)`, `TORCH_CHECK(input.size(i) == gradOutput_.size(i), ...)` with the upstream message text, placed next to the existing rank check.

## Ownership

The upstream PR touches only `aten/src/ATen/native/AdaptiveAveragePooling3d.cpp` and `aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu` plus a test; neither file is on the XPU dispatch path. No upstream issue or PR claims the XPU side, and no existing `intel/torch-xpu-ops` issue or PR covers this validation. Independent XPU work is required here.

## Environment

```
torch 2.15.0.dev20260830+xpu
python 3.13.13 (/opt/conda/bin/python3.13)
torch path /opt/conda/lib/python3.13/site-packages/torch/__init__.py
xpu device Intel(R) Arc(TM) Pro B60 Graphics
```

Filed by the automated XPU alignment pipeline after independent review of the scan evidence. The reproducer above was executed by the deterministic runner, not by the reviewing agent.

**Comment by laifenxiawucha** [comment]
@torchxpubot review

**Comment by torchxpubot** [comment]
`@torchxpubot review` can only be used on pull request comments.

**Comment by laifenxiawucha** [comment]
@torchxpubot fix

**Comment by torchxpubot** [comment]
Automated fix session for this issue. Stage reports are appended below as the pipeline advances.



## Issue Triage

| Field | Value |
|-------|-------|
| Reproduction missing | no |
| Scope | torch-xpu-ops |
| Dependencies | (none) |
| Handling | agent-fixable |

*Automated by issue-triage.*



## Root-cause Analysis

- **Issue type:** kernel/operator bug
- **Fix repo:** torch-xpu-ops
- **Analyzed at:** torch-xpu-ops@180f7f92
- **Root cause:** `adaptive_avg_pool3d_backward_kernel` (`src/ATen/native/xpu/sycl/AdaptiveAveragePooling3dKernels.cpp:588-595`) validates only that `input` and `gradOutput_` have equal rank, never that their leading (batch/channel) dims agree. It then derives every loop bound from `input` -- `sizeD` from `input.size(0)`/`input.size(1)`, `sizeB` from `input.size(0)`, and `totalZ = sizeB * sizeD * isizeT` (lines 610-642) -- while `AdaptiveAvgPool3dBackwardKernelFunctor` indexes `gradOutput` as `gradOutput_ + (d * osizeT_ + ostartT) * osizeH_ * osizeW_` with `d = i_plane / isizeT_` running over that input-derived plane count (lines 413-424). A `gradOutput` with a smaller batch or channel dim therefore gets read past its end, and the call returns an input-shaped tensor instead of raising. `_adaptive_avg_pool3d_backward` dispatches per backend (`CPU`/`CUDA`/`XPU`), so there is no shared validation layer to inherit the guard from; the upstream PR adds it to the CPU and CUDA templates only and does not touch the XPU path (pytorch/pytorch#195304 open).
- **Fix strategy:** In `adaptive_avg_pool3d_backward_kernel` (`src/ATen/native/xpu/sycl/AdaptiveAveragePooling3dKernels.cpp`), add a per-dimension size check for `i` in `[0, input.dim() - 3)` immediately after the existing rank check, mirroring the upstream hunk's message text (`": input and gradOutput_ must have the same size at dim ", i, ", but got ", ...`). This is the single choke point for both `adaptive_avg_pool3d_backward_xpu` and `adaptive_avg_pool3d_backward_out_xpu`, so one guard covers both entry points. Adding it to the kernel rather than the dispatch wrappers keeps it next to the rank check it complements.
- **Verdict:** IMPLEMENTING -- analysis produced a fixable root cause

*Automated by fix-root-cause.*



## Implement Result

- **Target repo:** torch-xpu-ops
- **Analyzed at:** torch-xpu-ops@180f7f92
- **Skip added:** no
- **Ready for verify:** yes

```diff
--- a/src/ATen/native/xpu/sycl/AdaptiveAveragePooling3dKernels.cpp
+++ b/src/ATen/native/xpu/sycl/AdaptiveAveragePooling3dKernels.cpp
@@ -593,6 +593,19 @@ void adaptive_avg_pool3d_backward_kernel(
       input.dim(),
       " for `gradOutput_` but got dimensions ",
       gradOutput_.dim());
+  // The loop bounds below come from `input` while `gradOutput` is indexed with
+  // them, so mismatched leading dims read past the end of `gradOutput`.
+  for (int64_t i = 0; i < input.dim() - 3; ++i) {
+    TORCH_CHECK(
+        input.size(i) == gradOutput_.size(i),
+        __func__,
+        ": input and gradOutput_ must have the same size at dim ",
+        i,
+        ", but got ",
+        input.size(i),
+        " and ",
+        gradOutput_.size(i));
+  }
 
   checkAllSameGPU(
       "adaptive_avg_pool3d_backward_xpu",
```

Skip-guard review (fresh-context subagent, `allow_skip=false`): **APPROVE** -- the diff adds only a `TORCH_CHECK` validation loop; no skips, xfails, tolerance loosening, seed pinning, deleted assertions, or swallowed exceptions.

*Automated by fix-implement.*



## Verify Result

- **Target repo:** torch-xpu-ops
- **Refined command:** `python3 repro.py` (the issue's three-case reproducer, made to exit non-zero when any XPU case returns instead of raising)
- **Verdict:** PASSED -- all three mismatch cases now raise on XPU with the upstream message text and correct dim indices
- **Lint:** clean (`lintrunner` on the changed file, no formatting fixes needed; `spin` has no config at this commit)

### Before / after

Before = nightly wheel `2.15.0.dev20260905+xpu`. After = source build `2.15.0a0+git02f2f62` (pytorch base `02f2f62`, `TORCH_XPU_ARCH_LIST=bmg,pvc`) with the fix compiled in.

| Test case | Before | After |
|-----------|--------|-------|
| 4D channel mismatch `input=(4,2,2,2)` `grad=(2,1,1,1)` | returned `(4, 2, 2, 2)` | raises `... must have the same size at dim 0, but got 4 and 2` |
| 5D batch mismatch `input=(3,2,2,2,2)` `grad=(1,2,1,1,1)` | returned `(3, 2, 2, 2, 2)` | raises `... must have the same size at dim 0, but got 3 and 1` |
| 5D channel mismatch `input=(1,4,2,2,2)` `grad=(1,2,1,1,1)` | returned `(1, 4, 2, 2, 2)` | raises `... must have the same size at dim 1, but got 4 and 2` |

### Regression checks

- Valid-shape `adaptive_avg_pool3d` forward and backward still match CPU bit-for-bit within tolerance across 4D and 5D inputs, including a non-divisible output size that takes the atomic path.
- Upstream `test/nn/test_pooling.py -k "adaptive_avg_pool and xpu"`: 6 passed.
- Wider `test/nn/test_pooling.py -k xpu`: 146 passed, 35 skipped, 1 failed -- `test_pool3d_large_size_int64_xpu`, a pre-existing fp16 `max_pool3d` numerics failure on a 2.24e9-element tensor. That test never calls `adaptive_avg_pool3d`, so it is unrelated to this change.

**Verification caveat:** the "before" is a nightly wheel and the "after" is a source build with the fix, so a second cold build without the fix was not run to isolate causation. In this case the message text in the "after" output is emitted by the added `TORCH_CHECK` itself, which ties the behavior change directly to this diff.

*Automated by fix-verify.*



## Summary

- **Issue:** intel/torch-xpu-ops#5232 -- `_adaptive_avg_pool3d_backward` on XPU accepts mismatched leading dims and reads gradOutput past its end
- **Path:** single-bug
- **Outcome:** IMPLEMENTING(fix_verified)
- **Branch:** `agent/fix-issue-5232` (base `180f7f92`, 1 commit, 1 file, +13)
- **Files changed:** `src/ATen/native/xpu/sycl/AdaptiveAveragePooling3dKernels.cpp`

`adaptive_avg_pool3d_backward_kernel` validated only rank equality between `input` and `gradOutput_`, then derived every loop bound from `input` while indexing `gradOutput` -- so a smaller batch or channel dim on `gradOutput` was read past its end and the op returned an input-shaped tensor instead of raising. The fix adds the per-dimension guard for `i` in `[0, input.dim() - 3)` next to the existing rank check, using the same message text as the upstream CPU/CUDA hunk.

Scope note: this covers the XPU path only. CPU and CUDA still lack the guard on trunk because pytorch/pytorch#195304 is still open; if its wording changes before merge, the XPU message should be re-checked against it. Once that PR lands, its device-generic `test_adaptive_avg_pooling_backward_fails` additions will exercise this guard on XPU automatically, so no XPU-side test was added here.

The fix is committed on the branch above; nothing was pushed and no PR was opened. The workflow exports `base_sha..branch` as a patch artifact for a human to apply.

_Generated by [fix job](https://github.com/intel/torch-xpu-ops/actions/runs/34106135873)._

**Comment by laifenxiawucha** [comment]
Closing as premature. Verified on an Intel Data Center GPU Max 1100 (`torch 2.12.1+xpu`): CPU accepts the same mismatched leading dims without raising, so this is a validation gap shared by all backends, not a CPU-vs-XPU divergence. The out-of-bounds read is real at the source level (loop bounds are derived from `input` while `gradOutput` is indexed with them) but was never directly demonstrated; an ASan-instrumented build would be needed.

The upstream contract is still unsettled: pytorch/pytorch#195304 is a **draft** with no review. Landing an XPU-only `TORCH_CHECK` now would make XPU the only backend rejecting these shapes, with error text that may not match what upstream settles on.

**This issue is not a canonical tracker.** If #195304 merges, the merge event re-enters the alignment candidate set and a fresh issue should be filed. Closed together with #5233, which is the same gap in the 2D kernel.

**Comment by laifenxiawucha** [comment]
**weak-rejected** 

why: filter-wrong
notes: `adaptive_avg_pool3d_backward_kernel` really does compare only
rank and then derive every loop bound from `input`, but CPU accepts the
same shapes, so the guard has to wait for pytorch/pytorch#195304 to
settle its wording; the patch also has no test at all.
